### PR TITLE
Label panes and other gadgets take face/size from *default-text-style*

### PR DESCRIPTION
**Pull request by john-a-carroll**
I decided against defaulting gadgets to :roman face - why force this choice on the user? Instead they pick up the face from `*default-text-style*`. The compose-space method on slider-pane needed a bit more care with the minor dimension size to properly take into account the width/height of the value. I also fixed typos in popup-init (`:suggest-` -> `:suggested-`) while I was editing that function.

### CODE REVIEW

File: Core/clim-core/gadgets.lisp [hunk 1]
```diff
@@ -520,7 +520,9 @@
                      :reader slider-number-of-quanta))
   (:documentation "The value is a real number, and default value for orientation is :vertical,
 and must never be nil.")
-  (:default-initargs :orientation :vertical))
+  (:default-initargs
+    :text-style (merge-text-styles (make-text-style :sans-serif nil nil) *default-text-style*)
```
Comment (scymtym) [inline Core/clim-core/gadgets.lisp]: Please add a function like `default-gadget-text-style' so this initform doesn't have to repeated as many times.

File: Core/clim-core/gadgets.lisp [hunk 2]
```diff
@@ -1659,13 +1661,22 @@ and must never be nil.")
...
-  (let ((minor (+ 8 (* 2 4) (if (gadget-show-value-p pane) 30 0)))
-        (major 128))
-    (if (eq (gadget-orientation pane) :vertical)
-        (make-space-requirement :min-width  minor :width  minor
-                                :min-height major :height major)
-        (make-space-requirement :min-width  major :width  major
-                                :min-height minor :height minor))))
+  (multiple-value-bind (vw vh)
+      (text-size pane (format-value (gadget-min-value pane) (slider-decimal-places pane)))
+    (setq vw
```
Comment (scymtym) [inline Core/clim-core/gadgets.lisp]: `alexandria:maxf` can be used here.

Why is `vh` not updated in the same way?

File: Core/clim-core/gadgets.lisp [hunk 3]
```diff
@@ -1659,13 +1661,22 @@ and must never be nil.")
...
+      (max vw
+           (text-size pane
+                      (format-value (gadget-max-value pane) (slider-decimal-places pane)))))
+    (let ((minor
+            (if (gadget-show-value-p pane)
+                (* 2 (+ 10.0 (if (eq (gadget-orientation pane) :vertical) vw vh))) ; the value
```
Comment (scymtym) [inline Core/clim-core/gadgets.lisp]: Please avoid (by defining constants) or explain the new magic numbers.

File: Core/clim-core/gadgets.lisp [hunk 4]
```diff
@@ -2624,10 +2639,9 @@ if INVOKE-CALLBACK is given."))
           ;;       which I could have easily worked around without adding kludges
           ;;       to the scroller-pane..
           (let* ((topmost-pane (if scroll-p
-                                  ;list-pane
                                   (scrolling (:scroll-bar :vertical
-                                              :suggest-height height   ;; Doesn't appear to be working..
-                                              :suggest-width (if scroll-p (+ 30 (bounding-rectangle-width list-pane))))
+                                              :suggested-height height
```
Comment (scymtym) [inline Core/clim-core/gadgets.lisp]: Please explain this change in the commit message or even make a separate commit.

File: Core/clim-core/gadgets.lisp (hunk 1 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: 1. `merge-text-styles` is unnecessary here. `make-text-style` will merge with `*default-text-style*` all NIL arguments.
2. shouldn't `(:default-initargs :text-style (make-text-style :sans-serif nil nil))` be supplied as a `basic-gadget` class option? I don't see a need for `default-gadget-text-style`, but maybe I miss something (?)

File: Core/clim-core/menu.lisp [hunk 5]
```diff
@@ -282,8 +282,8 @@ account, and create a list of menu buttons."
 
 ;;; Menu creation from command tables
 
-(defparameter *enabled-text-style*  (make-text-style :sans-serif :roman :normal))
-(defparameter *disabled-text-style* (make-text-style :sans-serif :roman :normal))
+(defparameter *enabled-text-style*  (make-text-style :sans-serif nil nil))
+(defparameter *disabled-text-style* (make-text-style :sans-serif nil nil))
```
Comment (dkochmanski) [inline Core/clim-core/menu.lisp]: since we have to create a text-style on each invocation *anyway* I think that we should inline them (and remove these special variables). Each of them is used only twice in the same function. We won't need merging text style then as well, because (make-text-style …) fills NIL's with the default one.

Comment (john-a-carroll) [reply Core/clim-core/menu.lisp]: My reasoning might be wrong, but I think the underlying issue is that while labels and other text associated with gadgets is output in the fully specified style eventually produced by medium-merged-text-style, the internal layout of gadgets (and compose-space etc) is guided by pane-text-style, which might have unspecified components. When the size is unspecified, methods such as text-style-ascent seem to take it as 12. To avoid this mismatch, my commits were intended to fully specify the pane text style from the outset. Another solution could be to simulate the outcome of medium-merged-text-style throughout gadgets.lisp - but that would require very many changes.

File: Core/clim-core/gadgets.lisp (hunk 4 above)
Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: The change is referred to in the commit message: fixed typos in popup-init (`:suggest-` -> :`suggested-`). The effect is so minor I didn't think it deserved a separate commit.

File: Core/clim-core/gadgets.lisp (hunk 1 above)
Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: Maybe this is the wrong solution. See my other reply about the underlying issue.

File: Core/clim-core/gadgets.lisp (hunk 2 above)
Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: Yes, `maxf` is better. Updating `vh` seems like overkill, since there's only a difference for fonts that have digits of varying heights, it only affects the bounding box for horizontal sliders, and in any case the effect is minimal.

File: Core/clim-core/gadgets.lisp (hunk 4 above)
Comment (scymtym) [reply Core/clim-core/gadgets.lisp]: I only saw that in the description of the pull request, not the commit message itself. It is generally useful to include all relevant information in the commit message.

File: Core/clim-core/gadgets.lisp (hunk 3 above)
Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: These values (2, 10.0, 8.0 and 1+) tie in with the slider-pane handle-repaint method, so yes, it would improve readability to make some of them constants and change handle-repaint accordingly.

Comment (scymtym) [reply Core/clim-core/gadgets.lisp]: Given that #647 contains some refactoring of the slider gadget, maybe it is fine to leave the magic numbers for now.

File: Core/clim-core/gadgets.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: is it costful though? if not, then having correct behavior is better.

File: Core/clim-core/gadgets.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: #647 has been merged.

File: Core/clim-core/menu.lisp (hunk 5 above)
Comment (dkochmanski) [reply Core/clim-core/menu.lisp]: Thank you for the explanation, I'll look into the code to have a better understanding of that.

File: Core/clim-core/gadgets.lisp (hunk 2 above)
Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: You're right. Commit https://github.com/McCLIM/McCLIM/pull/655/commits/a695dea5dc52727f1f6099f8fd622a3503981627 includes an improved `vw` and `vh` computation that is correct in all cases.

File: Core/clim-core/gadgets.lisp [hunk 6]
```diff
@@ -1659,8 +1662,19 @@ and must never be nil.")
...
+                        (format-value (gadget-min-value pane) (slider-decimal-places pane)))
+             (text-size pane
+                        (format-value (gadget-max-value pane) (slider-decimal-places pane)))))
+         (vh
+           (text-style-ascent (pane-text-style pane) pane))
+         (minor
+           (if (gadget-show-value-p pane)
+               (* 2 (+ 10.0 (if (eq (gadget-orientation pane) :vertical) vw vh))) ; the value
+               (* 2 (1+ 8.0)))) ; the knob
+         (major 128))
```
Comment (dkochmanski) [inline Core/clim-core/gadgets.lisp]: I think that this change (in general) operates on top of invalid assumption that min/max value is the narrowest/widest (especially if we agree, that text-style may be something else than `:fix`).

For instance, let's take range from 0 to 111 - in this case 100 is wider than both (assuming `0` is wider than `1`).
As of vertical panes ascent suggests, that font will never have numbers having pixels /below/ the baseline. I think it is a reasonable assumption, but point still stands.

I think that instead of having vw and vh values this should be rewritten to take into account the worst case scenario using provided clim primitives for that. I think that something in this spirit should be OK:

```
(let* ((value-size (etypecase (gadget-orientation pane)
                     (:vertical (text-style-height (pane-text-style pane) pane))
                     (:horizontal (let* ((dp (slider-decimal-places pane))
                                         (s1 (format-value (gadget-min-value pane) dp))
                                         (s2 (format-value (gadget-max-value pane) dp)))
                                    (* (max (length s1) (length s2))
                                       (text-style-width (pane-text-style pane) pane))))))
       (minor (if (gadget-show-value-p pane)
                  (* 2 (+ 10.0 value-size))
                  (* 2 (1+ 8.0))))
       (major 128))
...)
```

note that in this case compose-space will demand more space, because height is factually `(+ ascent descent)` and width is `(* number-of-characters size-of-letter-m)`. But IMHO this approach is more general than probing for text-size of min/max values.

Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: also I'd add better comments where these numeric constants come from. 10.0, 8.0, why 1+? why *2? all this is not obvious. I'll look into main part with text-styles for merging tomorrow.

Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: Thanks for your thoughtful suggestions. In my testing I hadn't noticed intermediate values being wider than min/max because DejaVu Sans (and it turns out many other fonts) have decimal digits that are all the same width. However, I believe the following deals correctly with variable width digits. It probes each digit individually since in some fonts even the widest digit is significantly narrower than the letter "m". For the other dimension, I used the font ascent not the height because draw-text* in handle-repaint draws with y alignment `:baseline`, not `:bottom`.
```
(defmethod compose-space ((pane slider-pane) &key width height)
  (declare (ignore width height))
  (let* ((value-size (ecase (gadget-orientation pane)
                       (:horizontal (text-style-ascent (pane-text-style pane) pane))
                       (:vertical (let* ((dp (slider-decimal-places pane))
                                         (s1 (format-value (gadget-min-value pane) dp))
                                         (s2 (format-value (gadget-max-value pane) dp)))
                                    (* (max (length s1) (length s2))
                                       (loop for c across "0123456789.-"
                                             maximize (text-size pane c)))))))
          (minor (if (gadget-show-value-p pane)
                     (* 2 (+ 10.0 value-size)) ; the value
                     (* 2 (1+ 8.0)))) ; the knob
          (major 128))
    (if (eq (gadget-orientation pane) :vertical)
        (make-space-requirement :min-width  minor :width  minor
                                :min-height major :height major)
        (make-space-requirement :min-width  major :width  major
                                :min-height minor :height minor))))
```
The 10.0 and 8.0 are used in the same way as in handle-repaint. The 1+ corresponds to the draw-circle* call that includes 1+ and 8.0. The 2 is because sliders are drawn with the same amount of space on one side of the line as the other. These values are more obvious when looked at in the context of the complete source code file.

Comment (john-a-carroll) [reply Core/clim-core/gadgets.lisp]: On further reflection, a McCLIM programmer would be ill-advised to use sliders with `:show-value-p t` in conjunction with a proportional font that has non-monospaced digits. If they do, they will get the unpleasant visual effect described in the article [Proportional vs. Monospaced Numbers: When to use which one in order to avoid "Wiggling Labels"](https://blog.usejournal.com/proportional-vs-monospaced-numbers-when-to-use-which-one-in-order-to-avoid-wiggling-labels-e31b1c83e4d0)

Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: thanks for the interesting link!  regardless, your approach with maximize and text-size looks good to me. ill advised or not, this is not our problem.

regarding magic values: my intention was to suggest, that you should put comment describing where they come from (i.e `magic values taken from handler-repaint`).

File: Core/clim-core/gadgets.lisp (hunk 2 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: this change is superseded. resolving this conversation.

File: Core/clim-core/gadgets.lisp (hunk 3 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: same concern is voiced against new code, resolving this conversation.

File: Core/clim-core/gadgets.lisp (hunk 4 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: please amend commit description, there is no need for a separate commit, but noting it in the commit message makes sense

**Comment by dkochmanski** [comment]
also it is worth noting, that pull request is related to this issue: https://github.com/McCLIM/McCLIM/issues/624

File: Core/clim-core/gadgets.lisp (hunk 6 above)
Comment (dkochmanski) [reply Core/clim-core/gadgets.lisp]: OK, so the problem originated in extensions/fonts module. Please see branch default-text-style-fixes for a commit which fixes issue #624. It also modifies text-style initarg to be put directly in basic-gadget (only once) -- second commit. Please rebase your remaining changes (`compose-space` fixes and `suggested-height` against that branch).

I'm not doing that myself because that would be kind of hijack of the pull request.

File: Core/clim-core/menu.lisp (hunk 5 above)
Comment (dkochmanski) [reply Core/clim-core/menu.lisp]: OK, so the problem originated in extensions/fonts module. Please see branch https://github.com/McCLIM/McCLIM/tree/default-text-style-fixes for a commit which fixes issue #624. It also modifies text-style initarg to be put directly in basic-gadget (only once) -- second commit. Please rebase your remaining changes (compose-space fixes and suggested-height against that branch).

I'm not doing that myself because that would be kind of hijack of the pull request. This fix is a direct result of peer reviewing your work, so all credit belongs to you.

**Comment by dkochmanski** [comment]
nb (regarding last push) - please rebase changes so there are no spuriuos "Merge branch" commits.

File: Core/clim-core/menu.lisp (hunk 5 above)
Comment (john-a-carroll) [reply Core/clim-core/menu.lisp]: Thanks for finding and fixing the problem! I don't need any credit, and I'm getting myself thoroughly confused in trying to rebase the remaining changes - probably due to the fact that I'm going down with flu. So please go ahead and make those changes yourself (as in commit https://github.com/McCLIM/McCLIM/pull/655/commits/1ace64b6facef6fe2ec50ed9f1412a2ed3820aba and changing :suggest- -> :suggested- x2)

**Comment by dkochmanski** [comment]
Alright, I'll go ahead then and make a new PR. Get healthy soon!
